iwyu_output.cc

```diff
@@ -507,6 +508,21 @@ OneIncludeOrForwardDeclareLine::OneIncludeOrForwardDeclareLine(
   end_linenum_ = GetLineNumber(GetInstantiationLoc(decl_lines.getEnd()));
 }
 
+OneIncludeOrForwardDeclareLine::OneIncludeOrForwardDeclareLine(
```

Comment:
Just checking - so this elaborated type line is treated the same as any other forward declaration for the type (it `matches()` the same), but if this one `is_desired()` we don't output anything, as it's not an explicit forward declaration.

From the perpective of what's needed in the code, this seems ok. From the perspective of the developer running IWYU, I feel like I'd expect to see some output for the type in question. i.e. 'ClassXX used - no forward declaration needed due to elaborated type on line XX-YY'

Reply:
I think it would not fit very well into "should add"/"should remove"/"full list" sections. As for warnings, IWYU doesn't emit any warning for the true fwd-declaration case too, if the fwd-decl appears before any fwd-decl-use in the same file. Elaborated type occurence is itself reported under verbosity level 6.

Reply:
There is some precedent for that with why-comments, e.g. `// for Type` which sometimes have a bit more context as to why a symbol is needed, e.g. `// for Type (for autocast)`. But these are set up at the point of reporting in iwyu.cc.

And I've actually considered removing them because they are a bit of hassle to maintain in traversal. Or potentially extending them to some kind of provenance (e.g. `std::vector<string> reasons` on `ASTNode`, so that traversal can push notes along the way and leave a breadcrumb trail up to where a use decision was made). But I'd really like not to add more noise in their current form.

Reply:
> And I've actually considered removing them

I can say that they are useful for me. Because I usually replace headers suggested for autocast or fn return to fwd-declarations. But I actually want to add command line flags to supersede default IWYU policies.

Reply:
> I think it would not fit very well into "should add"/"should remove"/"full list" sections.

Agreed. I was just thinking out loud what I'd hope to see.

> Or potentially extending them to some kind of provenance

Infrastructure to do this sounds interesting. Back to this commit, I've no issue with what you're currently doing here :)

Reply:
I'm actually a little surprised the whole `matches` machinery plays out correctly with pointer identity for `NamedDecl`'s. Do they all become the same decl in a scenario like this?
```
void f(class X*);  // decl=0x12345678?
void g(X*);        // decl=0x12345678?
class X;           // decl=0x12345678?
class X;           // decl=0x12345678?
```

Or maybe there's some canonicalization happening somewhere else.

Reply:
> Do they all become the same decl in a scenario like this?

The last two are actually redeclarations of the "owned" declaration from the first line, and they have their own addresses. Clang `TagType::getDecl()` function returns first declaration occured in a translation unit. `CalculateIwyuForForwardDeclareUse` can then remap a use to a forward declaration from the same file ([step D1](https://github.com/include-what-you-use/include-what-you-use/blob/0.20/iwyu_output.cc#L1034)).

Reply:
Ah, it was the D1 canonicalization I was missing, thanks!